【JavaScript 課題2】簡易Todoリストを作成しました。

solarsoyo opened the pull request:
コードレビューをお願いいたします。

Commit:
【JavaScript 課題2】簡易Todoリストを作成しました。
コードレビューをお願いいたします。

manaberuit commented on js/main.js:
```diff
@@ -0,0 +1,97 @@
+    const cell4 = row.insertCell(3);
+    cell1.textContent = key;
+    cell2.textContent = newTask;
+    cell3.innerHTML = `<input type="button" name="status-click" value="作業中">`;
```
HTMLを直で書くのはやめた方が良いですね。
`createElement`を用いた方が良いです。

solarsoyo replied:
ご指摘に基づき修正しました。
https://github.com/solarsoyo/javascript_practice2/pull/1/commits/00e76a9d8a321b99f329b833df605e3d9b6ca3b6

manaberuit commented on js/main.js:
```diff
@@ -0,0 +1,97 @@
+    //入力した追加フォームを空にする
+    document.getElementById('new-task').value = "";
+    changeStatus();
+    deletion();
```
なぜここで以下の２つの関数が呼び出されているのでしょう？？
```
changeStatus();
deletion();
```

solarsoyo replied:
changeStatusとdeletion関数は、addTask関数の中で作成されたイベントのため、addTask関数の中で呼び出さないとイベントが発火されないためです。
関数の中で作成されたイベントを、関数外（スコープ外）から発火させる方法があれば教えていただきたいです。

manaberuit replied:
>changeStatusとdeletion関数は、addTask関数の中で作成されたイベントのため

ん？そんなことないと思いますよ。
addTask関数外で宣言されているかと思います。

ただ、やりたいことは分かりましたので、後述のコメントのように変更していきましょう。

manaberuit commented on js/main.js:
```diff
@@ -0,0 +1,97 @@
+  function changeStatus() {
+    for (let i = 0; i < key; i++) {
+      const doneClick = document.getElementsByName('status-click')[i];
+      doneClick.onclick = ()=> {
```
`onclick`を使用したことに何か理由がありますか？
特になければ、ここも`addEventListener`を使用した方が良いかと思います。
他のとこも同様です。

solarsoyo replied:
addEventListenerでは意図した動きとならないためです。
onclickだとイベントを繰り返さないのですが、addEventListenerだとイベントを繰り返してしまいます。
例えばタスクを3つ登録した場合、2つ目のタスクのステータスを変更（作業中ボタン）しようとすると、作業中ボタン→完了ボタン→作業中ボタンと変遷し、ブラウザ上ではクリックしても変化しないように見えてしまうからです。
なお1つ目のタスクをクリックすると、作業中ボタン→完了ボタン→作業中ボタン→完了ボタンとなるため、疑似的には変化したように見えます。
いずれにせよ、addEventListenerでイベントを繰り返さない方法があれば、ご教授ください。

manaberuit replied:
なるほど。。
やりたいことは分かりましたので、後述のコメントのように変更していきましょう。

manaberuit reviewed (commented):
現在の実装方法ですと、タスクを削除した後でもIDがどんどん加算されていってしまうかと思います。
タスクが削除されたときには、IDが振り直されるように実装を変更してみてください。

また、空欄のまま追加ボタンが押下されたときには何かしらのエラーメッセージを表示させる処理も追加してみましょう。

また、いくつかコメントもしてありますので、そちらもご確認ください。

Commit:
HTMLを直で書くのをやめ、DOMでエレメントや属性を追加しました。

Commit:
空欄のまま追加ボタンを押された時は、入力欄にエラーメッセージを表示させるようにしました。

Commit:
タスクが削除されたら、IDを振り直すように修正

タスクが削除されたら、削除された後のIDを詰めるように修正しています。

manaberuit commented:
全体の設計を見直していきましょう。
現在は、setAtrributeなどを用いて属性を付与しているかと思いますが、ここらへんを変えていきましょう。

まず、
```
changeStatus();
deletion();
```
を変えていきましょう。
現在はgetElement・・・を使用して要素を取得し、そこに処理を追加するという関数になっているかと思いますが、すこし回りくどいやり方となっています。

これらの関数は、各々のボタンを生成する関数にしましょう。
戻り値で、生成したボタン（もちろん処理付き）を返すようにしてください。
この際には、createElement()関数を用いて、ボタン要素そのものを生成することになるかと思います。

そして、追加ボタン押下時の処理で、これらのボタンを生成する関数を呼び出して、返ってきた値（＝ボタン）をtable要素に追加していきましょう。

solarsoyo commented:
> 全体の設計を見直していきましょう。
> 現在は、setAtrributeなどを用いて属性を付与しているかと思いますが、ここらへんを変えていきましょう。
> 
> まず、
> 
> ```
> changeStatus();
> deletion();
> ```
> 
> を変えていきましょう。
> 現在はgetElement・・・を使用して要素を取得し、そこに処理を追加するという関数になっているかと思いますが、すこし回りくどいやり方となっています。
> 
> これらの関数は、各々のボタンを生成する関数にしましょう。
> 戻り値で、生成したボタン（もちろん処理付き）を返すようにしてください。
> この際には、createElement()関数を用いて、ボタン要素そのものを生成することになるかと思います。
> 
> そして、追加ボタン押下時の処理で、これらのボタンを生成する関数を呼び出して、返ってきた値（＝ボタン）をtable要素に追加していきましょう。

こちら

Commit:
関数の変更

ボタンの処理をする関数2つ（changeStatus、deletion）の代わりに、ボタンの生成と処理を兼ねる関数を2つ作成しました。

manaberuit commented on js/main.js:
```diff
@@ -28,47 +28,41 @@
-    }
+  // 状態ボタン（最初は作業中ボタン）を作成する関数
+  function addStatusButton(cell) {
+    const statusButton = document.createElement('input');
```
inputタグでボタンを生成するやり方は古いやり方です。
https://developer.mozilla.org/ja/docs/Web/HTML/Element/Input/button
（メモを参照）

直接buttonタグを生成してください。

solarsoyo replied:
inputタグをbuttonタグで生成し、合わせてvalue属性でタスクの状態を設定していたのを、textContentで行うようにしました。
これまで、buttonタグがあるのに、なぜ教本などではinputタグを使うのだろうと疑問に思っていました。これからはbuttonタグを使うようにします。
fb2eb3ad0f52065fe62e13faf6b64f18eb660493

manaberuit commented on js/main.js:
```diff
@@ -28,47 +28,41 @@
+  // 状態ボタン（最初は作業中ボタン）を作成する関数
+  function addStatusButton(cell) {
+    const statusButton = document.createElement('input');
+    statusButton.setAttribute('type','button');
```
setAttributeを使用するのは避けましょう。
状況によって動作が変わってきてしまうためです。
プロパティを使用した方がよいです。
https://developer.mozilla.org/ja/docs/Web/API/Element/setAttribute
（注記参照）
https://www.softel.co.jp/blogs/tech/archives/1595

solarsoyo replied:
setAttributeで設定するのがスマートなのかと思っていました…。アドバイスに基づき、プロパティで対応しました。
なお、28行目はプロパティで対応できないためdata属性を設定しました。
※cell1.name =''では設定できないため
7cf24ab226e930191c1ed5e880bb916b4ffc6a48

見直した後、取得するdata属性のnameを指定した方が良いと思い、68行目を変更しました。
a233dbf7f419dcf56550f7fca074a12a6c8aaca2

manaberuit commented on js/main.js:
```diff
@@ -28,47 +28,41 @@
+    cell.appendChild(deletionButton);
+    // 削除ボタンをクリックしたときのイベント：親要素の親要素のtrタグを削除する
+    deletionButton.addEventListener('click', ()=> {
+      tbody.removeChild(deletionButton.parentNode.parentNode);
```
>parentNodeを繰り返し使っているところがあります。
親要素の親要素を取得するためにやったのですが、スマートなやり方があればご教授いただきたく存じます。

私の知る限り、こう記述するしかないですね。。

solarsoyo replied:
ありがとうございました。

manaberuit reviewed (commented):
いくつかコメントしましたので、ご確認ください。

Commit:
inputタグをbuttonタグに変更

buttonタグへの変更とあわせて、value属性をtextContentで設定や取得するように変更しました。

Commit:
setAttributeの代わりにプロパティ等を使用

28行目はプロパティを使用できないため、代わりにdata属性で対応しました。
42行目はプロパティで対応。
68行目はdata属性で対応したため、data属性を取得する命令文に変更しました。

Commit:
68行目のdata属性取得の命令文にname:IDを指定するようにしました。

manaberuit commented:
OKです！！
マージしてしまってください^^
